File: 32.最长有效括号.py
#
# @lc app=leetcode.cn id=32 lang=python
#
# [32] 最长有效括号
#

# 思路：用一个栈记录括号位置，每匹配成功一个括号，就把左右括号出现的位置放到一个数组里。
# 数组排序以后，他的最大连续区间就是最长有效括号串。
# @lc code=start
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        st, bitmap = [], [0] * len(s)
        for i, c in enumerate(s):
            if c == '(':
                st.append(i)
            elif st:
                bitmap[i], bitmap[st.pop()] = 1, 1

        res, tmp = 0, 0
        for i in bitmap:
            if i == 1:
                tmp += 1
            else:
                res = max(tmp, res)
                tmp = 0
        return max(res, tmp)
# @lc code=end

File: 32.最长有效括号.cpp
/*
 * @lc app=leetcode.cn id=32 lang=cpp
 *
 * [32] 最长有效括号
 */

// @lc code=start
class Solution {
public:
    int longestValidParentheses(string s) {
        if (s.size() < 2) {
            return 0;
        }
        std::vector<int> v(s.size(), 0);
        stack<int> t;
        for (int i = 0; i < s.size(); ++i) {
            if (s[i] == ')') {
                if (!t.empty()) {
                    v[i] = 1;
                    v[t.top()] = 1;
                    t.pop();
                }
            } else {
                t.push(i);
            }
        }
        int res = 0;
        int tmp = 0;
        for (int i = 0; i < v.size(); ++i) {
            if (v[i] == 1) {
                tmp++;
            } else {
                res = max(res, tmp);
                tmp = 0;
            }
        }
        return max(res, tmp);
    }
};
// @lc code=end

